twoliter/src/project/migrate/migrator.rs

```diff
@@ -0,0 +1,409 @@
+    pub fn register(&mut self, key: (u32, u32), migrator: Box<dyn ProjectMigrator>) {
+        self.migrators.insert(key, migrator);
+    }
+
+    /// Find a migrator that can handle the given version transition.
+    fn find_migration_edge(
+        &self,
+        from_version: u32,
+        to_version: u32,
+    ) -> Option<&dyn ProjectMigrator> {
+        self.migrators
+            .get(&(from_version, to_version))
+            .map(|m| m.as_ref())
+    }
+
+    /// Find the shortest migration path from source version to target version using BFS
+    fn find_migration_path(&self, source: u32, target: u32) -> Option<Vec<(u32, u32)>> {
+        if source == target {
+            return Some(Vec::new());
+        }
+
+        let mut graph: HashMap<u32, Vec<u32>> = HashMap::new();
+        for &(src, dst) in self.migrators.keys() {
+            graph.entry(src).or_default().push(dst);
+        }
+
+        let mut queue = VecDeque::new();
+        let mut visited = HashSet::new();
+        let mut parent: HashMap<u32, u32> = HashMap::new();
+
+        queue.push_back(source);
+        visited.insert(source);
+
+        while let Some(current) = queue.pop_front() {
+            if current == target {
+                return Some(self.reconstruct_path(&parent, source, target));
+            }
+
+            if let Some(neighbors) = graph.get(&current) {
+                for &next in neighbors {
+                    if !visited.contains(&next) {
+                        visited.insert(next);
+                        parent.insert(next, current);
+                        queue.push_back(next);
+                    }
+                }
+            }
+        }
+        None
+    }
+
+    /// Reconstructs the path from source to target using the parent map
+    fn reconstruct_path(
+        &self,
+        parent: &HashMap<u32, u32>,
+        source: u32,
+        target: u32,
+    ) -> Vec<(u32, u32)> {
+        let mut path = Vec::new();
+        let mut current = target;
+
+        while current != source {
+            let prev = *parent.get(&current).unwrap(); // Safe because we know the path exists
+            path.push((prev, current));
+            current = prev;
+        }
+
+        path.reverse();
+        path
+    }
+
+    /// Migrate a project file from a version to the next version.
+    ///
+    /// This method will find the appropriate migration path and apply all necessary
+    /// transformations to get from the source version to the target version.
+    pub fn migrate_project(
+        &self,
+        mut content: Box<dyn Any>,
+        source: u32,
+        target: u32,
+    ) -> Result<Box<dyn Any>> {
+        if source == target {
+            return Ok(content);
+        }
+
+        let path = self.find_migration_path(source, target).ok_or_else(|| {
+            anyhow!(
+                "Failed to find migration path from v{} to v{}",
+                source,
+                target
+            )
+        })?;
```

Comment:
nit: the `context` methods are the idiomatic way to do this.

```suggestion
        let path = self.find_migration_path(source, target).with_context(||
            format!("Failed to find migration path from v{source} to v{target}")
        )?;
```